### PR TITLE
Fix #4148: Do not emit static forwarders for non-top-level objects by default.

### LINKED ISSUE #4148: Unable to find extant inner trait IR during fastopt
Please refer to this tag: https://github.com/djspiewak/cats-effect/tree/bug/scalajs-concurrentops

You can reproduce the issue with `sbt coreJS/test`. `lawsJS/test` also reproduces it. ScalaJS 1.1.1, Scala 2.13.3/2.12.11 (reproduces on both). Compilation finishes without a problem, but fastOpt does not:

```
[error] Referring to non-existent class cats.effect.Concurrent$Ops
[error]   called from private cats.effect.Concurrent$.$anonfun$continual$1(java.lang.Object,cats.effect.Concurrent,scala.Function1,cats.effect.concurrent.Deferred)java.lang.Object
[error]   called from cats.effect.Concurrent$.continual(java.lang.Object,scala.Function1,cats.effect.Concurrent)java.lang.Object
[error]   called from cats.effect.Concurrent.continual(java.lang.Object,scala.Function1)java.lang.Object
[error]   called from private cats.effect.internals.MVarConcurrent.$anonfun$modify$3(scala.Function1,cats.effect.concurrent.Ref,java.lang.Object)java.lang.Object
[error]   called from private cats.effect.internals.MVarConcurrent.$anonfun$modify$1(scala.Function1,cats.effect.concurrent.Ref)java.lang.Object
[error]   called from cats.effect.internals.MVarConcurrent.modify(scala.Function1)java.lang.Object
[error]   called from private cats.effect.concurrent.BaseMVarTests.$anonfun$new$212(cats.effect.concurrent.MVar2,int)cats.effect.IO
[error]   called from private cats.effect.concurrent.BaseMVarTests.$anonfun$new$211(cats.effect.concurrent.MVar2,java.lang.Void)cats.effect.IO
[error]   called from private cats.effect.concurrent.BaseMVarTests.$anonfun$new$210(cats.effect.concurrent.MVar2)cats.effect.IO
[error]   called from private cats.effect.concurrent.BaseMVarTests.$anonfun$new$209()scala.concurrent.Future
[error]   called from constructor cats.effect.concurrent.BaseMVarTests.<init>()void
[error]   called from constructor cats.effect.concurrent.MVarAsyncTests.<init>()void
[error]   called from static constructor cats.effect.concurrent.MVarAsyncTests.<clinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   cats.effect.Concurrent$
[error]   cats.effect.internals.MVarConcurrent
```

There's a bunch of that, but this is the topmost one. The perplexing thing is that the `Concurrent$Ops.sjsir` file *exists*, and `coreJVM/test` works just fine. The error survives through `clean` and even `rm -rf *; git reset --hard HEAD`.

Note that this source code used to use Simulacrum, but I removed that on this tag. You can see the pre-removal state in `HEAD^` on the tag. The source code *itself* is identical to the **series/2.x** branch on the same repository, which builds and tests just fine with ScalaJS 1.1.1.

I'm going to keep poking around with the plugins to see if I can figure out where the error was introduced, but this *seems* like an SJS bug.

### PR DESCRIPTION
**Pull request by sjrd**
Alternative to #4149.

---

### Fix #4148: Do not emit static forwarders for non-top-level objects by default.

Scala/JVM does not emit static forwarders for non-top-level objects. The reason we do it in Scala.js is for implementations of JDK classes that contain static methods in inner static classes (#3950). However, this causes issues with existing Scala projects that have non-top-level `object`s and `class`es with names differing only in case.

In this commit, we now only emit static forwarders for top-level objects by default, and add an opt-in option to emit them for non-top-level objects, which should be used by implementations of JDK classes.

---

### Do not erase a regular class with a static forwarder class.

This is only relevant for case insensitive file systems. On such file systems, a static nested object could generate a static forwarder class whose name differed from a regular class' name only in case, therefore erasing the latter on case insensitive file systems.

We now avoid emitting static forwarder classes for objects if they would clash with a regular class on a case insensitive file system.

When doing so, we emit a warning. On Scala 2.13.2+, the warning can be silenced with `@nowarn("other")` on the object, or `-Wconf:cat=other:s` in the scalac options.

For top-level objects, the warning is similar to a warning already emitted by the JVM back-end. For non-top-level objects, it only happens in Scala.js, and only when they are emitted due to the opt-in option.

This commit complements its parent in fixing #4148 for good, as now that issue won't present itself even when we opt in for static forwards of non-top-level objects. We therefore add a test for that issue, which we could not do before because the test suite as a whole requires the opt-in for other reasons.

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -1026,6 +1026,8 @@ object Build {
        * flag prevents these mistakes from happening.
        */
       scalacOptions += "-Yno-predef",
+      // We also opt in to emitting static forwarders for non-top-level objects
```
Comment (gzm0) [inline project/Build.scala:1029]: IMO this comment is not helpful. It describes *what* the code does not *why* it does it. I think a comment explaining why is warranted here.

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala [hunk 2]
```diff
@@ -364,14 +365,62 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
...
+           * systems.
+           */
+
+          /* I could not find any reference anywhere about what locale is used
+           * by case insensitive file systems to compare case-insensitively.
+           * In doubt, force the English locale, which is probably going to do
+           * the right thing in virtually all cases (especially if users stick
+           * to ASCII class names), and it has the merit of being deterministic,
+           * as opposed to using the OS' default locale.
+           * The JVM performs a similar test to emit a warning for conflicting
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:387]: JVM *backend*?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RegressionTest.scala [hunk 3]
```diff
@@ -906,4 +914,24 @@ object RegressionTest {
   object `class` { // scalastyle:ignore
     def foo(x: Int): Int = x + 1
   }
+
+  /* The objects and classes here intentionally have names that differ
+   * only in case, and are intentionally defined in a specific order.
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RegressionTest.scala:919]: Consider commenting *why* this is intentional.

**Comment by sjrd** [comment]
Updated.

**Pull request merged**